File: python/nemo_relay/providers/oci_genai.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Request and response codecs for Oracle Cloud Infrastructure (OCI) Generative AI.

The OCI Generative AI chat API wraps requests in a ``ChatDetails`` envelope
(``compartmentId``, ``servingMode``, ``chatRequest``) and supports two chat
request formats selected by ``chatRequest.apiFormat``:

- ``GENERIC``: OpenAI-style ``messages`` whose ``content`` is a list of typed
  parts (``{"type": "TEXT", "text": ...}``). Used by Meta Llama, Google,
  xAI, OpenAI, and imported open-weights models (for example NVIDIA Nemotron)
  hosted on dedicated AI clusters.
- ``COHERE``: a single ``message`` string plus ``chatHistory`` turns with
  ``USER``/``CHATBOT``/``SYSTEM`` roles. Used by Cohere Command models.

``OCIGenAIChatCodec`` normalizes both formats into ``AnnotatedLLMRequest`` so
request intercepts (PII redaction, guardrails, policy) can operate on a single
shape, and merges intercept edits back without dropping provider-specific
fields. ``OCIGenAIResponseCodec`` normalizes ``ChatResult`` responses for
``LLMEnd`` event annotation.

Example::

    import nemo_relay
    from nemo_relay.providers.oci_genai import OCIGenAIChatCodec, OCIGenAIResponseCodec

    result = await nemo_relay.llm.execute(
        "oci-genai",
        nemo_relay.LLMRequest({}, chat_details),
        call_oci_chat,
        codec=OCIGenAIChatCodec(),
        response_codec=OCIGenAIResponseCodec(),
    )
"""

from __future__ import annotations

import typing

from nemo_relay import Json
from nemo_relay._native import AnnotatedLLMRequest, AnnotatedLLMResponse, LLMRequest

# OCI parameter name -> normalized GenerationParams field. Parameters without
# a normalized slot (topK, penalties, seed, ...) stay untouched in the raw
# payload, which encode() preserves.
_GENERIC_PARAM_MAP = {"maxTokens": "max_tokens", "temperature": "temperature", "topP": "top_p", "stop": "stop"}
_COHERE_PARAM_MAP = {
    "maxTokens": "max_tokens",
    "temperature": "temperature",
    "topP": "top_p",
    "stopSequences": "stop",
}

_COHERE_ROLE_TO_NORMALIZED = {"USER": "user", "CHATBOT": "assistant", "SYSTEM": "system", "TOOL": "tool"}
_NORMALIZED_ROLE_TO_COHERE = {v: k for k, v in _COHERE_ROLE_TO_NORMALIZED.items()}


def _get_first(mapping: typing.Mapping[str, typing.Any], *keys: str) -> typing.Any:
    """Return the first present key from ``mapping`` across naming conventions.

    The OCI SDKs emit camelCase JSON while the OCI CLI emits kebab-case.
    Callers pass camelCase keys; kebab-case and snake_case fallbacks are
    derived automatically.
    """
    for key in keys:
        for candidate in (key, _camel_to_kebab(key), _camel_to_snake(key)):
            if candidate in mapping:
                return mapping[candidate]
    return None


def _camel_to_kebab(key: str) -> str:
    return "".join(f"-{c.lower()}" if c.isupper() else c for c in key)


def _camel_to_snake(key: str) -> str:
    return "".join(f"_{c.lower()}" if c.isupper() else c for c in key)


def _generic_content_to_text(content: typing.Any) -> typing.Any:
    """Flatten a GENERIC content-part list into plain text when possible."""
    if isinstance(content, str) or content is None:
        return content
    if isinstance(content, list):
        texts: list[str] = []
        for part in content:
            if isinstance(part, dict) and _get_first(part, "type") == "TEXT":
                texts.append(str(_get_first(part, "text") or ""))
            else:
                # Non-text part (image, etc.): leave the list untouched.
                return content
        return "".join(texts)
    return content


def _text_to_generic_content(content: typing.Any) -> typing.Any:
    """Wrap plain text back into the GENERIC typed content-part list."""
    if isinstance(content, str):
        return [{"type": "TEXT", "text": content}]
    return content


def _normalize_usage(usage: typing.Any) -> dict[str, typing.Any] | None:
    """Map OCI usage counters onto the normalized ``Usage`` field names."""
    if not isinstance(usage, dict):
        return None
    normalized: dict[str, typing.Any] = {}
    prompt_tokens = _get_first(usage, "promptTokens")
    completion_tokens = _get_first(usage, "completionTokens")
    total_tokens = _get_first(usage, "totalTokens")
    if prompt_tokens is not None:
        normalized["prompt_tokens"] = prompt_tokens
    if completion_tokens is not None:
        normalized["completion_tokens"] = completion_tokens
    if total_tokens is not None:
        normalized["total_tokens"] = total_tokens
    return normalized or None


def _oci_tool_call_to_normalized(tool_call: typing.Any) -> typing.Any:
    """Convert a flat OCI ``toolCalls`` entry into the normalized nested shape."""
    if not isinstance(tool_call, dict) or "function" in tool_call:
        return tool_call
    return {
        "id": _get_first(tool_call, "id"),
        "type": "function",
        "function": {
            "name": _get_first(tool_call, "name"),
            "arguments": _get_first(tool_call, "arguments"),
        },
    }


def _normalized_tool_call_to_oci(tool_call: typing.Any) -> typing.Any:
    """Convert a normalized nested tool call back into the flat OCI shape."""
    if not isinstance(tool_call, dict) or "function" not in tool_call:
        return tool_call
    function = tool_call.get("function") or {}
    return {
        "id": tool_call.get("id"),
        "type": "FUNCTION",
        "name": function.get("name"),
        "arguments": function.get("arguments"),
    }


class OCIGenAIChatCodec:
    """Request codec for the OCI Generative AI chat API.

    Accepts either a full ``ChatDetails`` envelope or a bare ``chatRequest``
    payload, in the ``GENERIC`` or ``COHERE`` API format. Provider-specific
    fields that have no normalized equivalent are carried in
    ``AnnotatedLLMRequest.extra`` and restored by ``encode()``.
    """

    def decode(self, request: LLMRequest) -> AnnotatedLLMRequest:
        """Decode an OCI chat payload into ``AnnotatedLLMRequest``."""
        content = dict(request.content)
        chat_request = _get_first(content, "chatRequest")
        envelope = content if isinstance(chat_request, dict) else None
        chat_request = dict(chat_request) if isinstance(chat_request, dict) else content

        api_format = str(_get_first(chat_request, "apiFormat") or "GENERIC").upper()
        model = self._model_from_envelope(envelope)

        if api_format == "COHERE":
            messages, params = self._decode_cohere(chat_request)
        else:
            messages, params = self._decode_generic(chat_request)

        tools = _get_first(chat_request, "tools")
        tool_choice = _get_first(chat_request, "toolChoice")

        return AnnotatedLLMRequest(
            messages,
            model=model,
            params=params or None,
            tools=tools if isinstance(tools, list) else None,
            tool_choice=tool_choice,
            api_specific={"api": "custom", "api_name": "oci_genai", "data": {"apiFormat": api_format}},
            extra={"envelope": envelope, "chatRequest": chat_request},
        )

    def encode(self, annotated: AnnotatedLLMRequest, original: LLMRequest) -> LLMRequest:
        """Merge annotated edits back into the original OCI payload.

        Edits are applied by comparing the annotation against a decoded
        baseline of the original request and patching only changed fields, so
        ``encode(decode(original), original)`` returns the original payload at
        the JSON-value level and unmodeled provider fields survive edits.
        """
        content = dict(original.content)
        chat_request_value = _get_first(content, "chatRequest")
        has_envelope = isinstance(chat_request_value, dict)
        chat_request = dict(chat_request_value) if has_envelope else dict(content)

        api_specific = annotated.api_specific or {}
        api_data_value = api_specific.get("data")
        api_data: dict[str, typing.Any] = api_data_value if isinstance(api_data_value, dict) else {}
        api_format = str(api_data.get("apiFormat") or _get_first(chat_request, "apiFormat") or "GENERIC").upper()

        baseline = self.decode(original)
        edited_messages = [dict(message) for message in annotated.messages]
        baseline_messages = [dict(message) for message in baseline.messages]

        if edited_messages != baseline_messages:
            if api_format == "COHERE":
                self._encode_cohere(chat_request, annotated)
            else:
                self._patch_generic_messages(chat_request, edited_messages, baseline_messages)

        edited_params = dict(annotated.params or {})
        baseline_params = dict(baseline.params or {})
        if edited_params != baseline_params:
            param_map = _COHERE_PARAM_MAP if api_format == "COHERE" else _GENERIC_PARAM_MAP
            normalized_to_oci = {v: k for k, v in param_map.items()}
            for key, value in edited_params.items():
                if value is None or baseline_params.get(key) == value:
                    continue
                chat_request[normalized_to_oci.get(key, key)] = value

        if has_envelope:
            content["chatRequest"] = chat_request
            return LLMRequest(original.headers, content)
        return LLMRequest(original.headers, chat_request)

    @classmethod
    def _patch_generic_messages(
        cls,
        chat_request: dict[str, typing.Any],
        edited_messages: list[dict[str, typing.Any]],
        baseline_messages: list[dict[str, typing.Any]],
    ) -> None:
        """Rewrite only the GENERIC messages that intercepts actually changed.

        Unchanged messages are carried over from the raw payload verbatim so
        per-message provider fields without a normalized equivalent survive.
        """
        raw_messages = _get_first(chat_request, "messages") or []
        patched: list[typing.Any] = []
        for index, edited in enumerate(edited_messages):
            unchanged = index < len(baseline_messages) and edited == baseline_messages[index]
            if unchanged and index < len(raw_messages):
                patched.append(raw_messages[index])
            else:
                patched.append(cls._build_generic_message(edited))
        chat_request["messages"] = patched

    @staticmethod
    def _build_generic_message(message: typing.Mapping[str, typing.Any]) -> dict[str, typing.Any]:
        encoded: dict[str, typing.Any] = {
            "role": str(message.get("role") or "user").upper(),
            "content": _text_to_generic_content(message.get("content")),
        }
        if "tool_calls" in message:
            encoded["toolCalls"] = [_normalized_tool_call_to_oci(tc) for tc in message["tool_calls"]]
        if "tool_call_id" in message:
            encoded["toolCallId"] = message["tool_call_id"]
        return encoded

    @staticmethod
    def _model_from_envelope(envelope: typing.Mapping[str, typing.Any] | None) -> str | None:
        if not envelope:
            return None
        serving_mode = _get_first(envelope, "servingMode")
        if not isinstance(serving_mode, dict):
            return None
        model = _get_first(serving_mode, "modelId") or _get_first(serving_mode, "endpointId")
        return str(model) if model is not None else None

    @staticmethod
    def _decode_generic(
        chat_request: typing.Mapping[str, typing.Any],
    ) -> tuple[list[dict[str, typing.Any]], dict[str, typing.Any]]:
        messages: list[dict[str, typing.Any]] = []
        for message in _get_first(chat_request, "messages") or []:
            role = str(_get_first(message, "role") or "USER").lower()
            normalized: dict[str, typing.Any] = {
                "role": role,
                "content": _generic_content_to_text(_get_first(message, "content")),
            }
            tool_calls = _get_first(message, "toolCalls")
            if tool_calls is not None:
                normalized["tool_calls"] = [_oci_tool_call_to_normalized(tc) for tc in tool_calls]
            tool_call_id = _get_first(message, "toolCallId")
            if tool_call_id is not None:
                normalized["tool_call_id"] = tool_call_id
            messages.append(normalized)

        params = {
            normalized_key: chat_request[oci_key]
            for oci_key, normalized_key in _GENERIC_PARAM_MAP.items()
            if oci_key in chat_request
        }
        return messages, params

    @staticmethod
    def _decode_cohere(
        chat_request: typing.Mapping[str, typing.Any],
    ) -> tuple[list[dict[str, typing.Any]], dict[str, typing.Any]]:
        messages: list[dict[str, typing.Any]] = []

        preamble = _get_first(chat_request, "preambleOverride")
        if preamble:
            messages.append({"role": "system", "content": preamble})

        for turn in _get_first(chat_request, "chatHistory") or []:
            role = _COHERE_ROLE_TO_NORMALIZED.get(str(_get_first(turn, "role") or "USER").upper(), "user")
            messages.append({"role": role, "content": _get_first(turn, "message")})

        current = _get_first(chat_request, "message")
        if current is not None:
            messages.append({"role": "user", "content": current})

        params = {
            normalized_key: chat_request[oci_key]
            for oci_key, normalized_key in _COHERE_PARAM_MAP.items()
            if oci_key in chat_request
        }
        return messages, params

    @staticmethod
    def _encode_cohere(chat_request: dict[str, typing.Any], annotated: AnnotatedLLMRequest) -> None:
        messages = list(annotated.messages)

        if messages and str(messages[0].get("role")) == "system":
            chat_request["preambleOverride"] = messages[0].get("content")
            messages = messages[1:]

        current_message = ""
        if messages and str(messages[-1].get("role")) == "user":
            current_message = messages[-1].get("content") or ""
            messages = messages[:-1]

        history = [
            {
                "role": _NORMALIZED_ROLE_TO_COHERE.get(str(turn.get("role")), "USER"),
                "message": turn.get("content"),
            }
            for turn in messages
        ]

        chat_request["message"] = current_message
        if history or "chatHistory" in chat_request:
            chat_request["chatHistory"] = history


class OCIGenAIResponseCodec:
    """Response codec for OCI Generative AI ``ChatResult`` payloads.

    Normalizes both ``GENERIC`` (``choices``-based) and ``COHERE``
    (``text``-based) chat responses for ``LLMEnd`` event annotation. Accepts
    the full ``ChatResult`` (``modelId``, ``chatResponse``) or a bare chat
    response object.
    """

    def decode_response(self, response: Json) -> AnnotatedLLMResponse:
        """Decode a raw OCI chat response into ``AnnotatedLLMResponse``."""
        if not isinstance(response, dict):
            return AnnotatedLLMResponse(extra={"raw": response})

        chat_response = _get_first(response, "chatResponse")
        envelope = response if isinstance(chat_response, dict) else None
        chat_response = dict(chat_response) if isinstance(chat_response, dict) else dict(response)

        model = None
        if envelope is not None:
            model_value = _get_first(envelope, "modelId")
            model = str(model_value) if model_value is not None else None

        api_format = str(_get_first(chat_response, "apiFormat") or "GENERIC").upper()

        message: typing.Any = None
        tool_calls: typing.Any = None
        finish_reason: typing.Any = None

        if api_format == "COHERE":
            message = _get_first(chat_response, "text")
            tool_calls = _get_first(chat_response, "toolCalls")
            finish_reason = _get_first(chat_response, "finishReason")
        else:
            choices = _get_first(chat_response, "choices") or []
            if choices:
                first_choice = choices[0]
                finish_reason = _get_first(first_choice, "finishReason")
                raw_message = _get_first(first_choice, "message")
                if isinstance(raw_message, dict):
                    message = _generic_content_to_text(_get_first(raw_message, "content"))
                    tool_calls = _get_first(raw_message, "toolCalls")

        usage = _normalize_usage(_get_first(chat_response, "usage"))

        return AnnotatedLLMResponse(
            model=model,
            message=message,
            tool_calls=tool_calls if isinstance(tool_calls, list) else None,
            finish_reason=finish_reason,
            usage=usage,
            api_specific={"api": "custom", "api_name": "oci_genai", "data": {"apiFormat": api_format}},
        )


__all__ = [
    "OCIGenAIChatCodec",
    "OCIGenAIResponseCodec",
]

File: crates/core/src/codec/oci_genai.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Built-in codec for the Oracle Cloud Infrastructure (OCI) Generative AI chat API.
//!
//! Implements [`LlmResponseCodec`] (response decode) for the OCI Generative AI
//! chat format.
//!
//! # OCI-specific patterns handled
//!
//! - **Three API formats** selected by the response `apiFormat`: `GENERIC`
//!   (`choices`-based, OpenAI-style), `COHERE` (`text`-based), and `COHEREV2`
//!   (single `message` with typed content parts and nested `function` tool
//!   calls, per the OCI `CohereChatResponseV2` schema).
//! - **Responses**: `ChatResult` payloads (`modelId`, `chatResponse`); `usage`
//!   counters are `promptTokens`/`completionTokens`/`totalTokens`.
//! - **Unmodeled fields are preserved**: envelope and chat-response fields the
//!   normalized shape does not model (`timeCreated`, future provider fields)
//!   are carried in `extra` rather than discarded, consistent with the other
//!   response codecs. Unmodeled fields of the decoded choice and assistant
//!   message — `logprobs`, `serviceTier`, `groundingMetadata`,
//!   `reasoningContent`, `refusal` (GENERIC) and `toolPlan`, `citations`
//!   (COHEREV2) per the OCI schema — are namespaced in `extra` under
//!   `"choice"` and `"message"`.
//!
//! The codec accepts the REST wire format only: camelCase keys, as documented
//! in the OCI API reference. Alternate renderings produced by Oracle tooling
//! (the CLI's kebab-case `data` envelope, `oci.util.to_dict()` snake_case
//! dicts) are the caller's responsibility to convert.

use crate::error::{FlowError, Result};
use crate::json::Json;

use super::request::{ContentPart, MessageContent, ProviderNativeComponent};
use super::response::{
    AnnotatedLlmResponse, ApiSpecificResponse, FinishReason, ResponseToolCall, Usage,
};
use super::traits::LlmResponseCodec;

// ---------------------------------------------------------------------------
// Public codec struct
// ---------------------------------------------------------------------------

/// Built-in codec for the OCI Generative AI chat API.
pub struct OCIGenAIChatCodec;

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

/// Map an OCI finish reason string to normalized [`FinishReason`].
///
/// GENERIC responses use OpenAI-style lowercase reasons (Gemini models emit
/// `max_tokens` for the length stop); COHERE and COHEREV2 responses use
/// UPPERCASE Cohere reasons (`TOOL_CALL` and `STOP_SEQUENCE` are V2-only).
fn map_oci_finish_reason(reason: &str) -> FinishReason {
    match reason {
        "stop" | "COMPLETE" | "STOP_SEQUENCE" => FinishReason::Complete,
        "length" | "max_tokens" | "MAX_TOKENS" => FinishReason::Length,
        "tool_calls" | "TOOL_CALL" => FinishReason::ToolUse,
        "content_filter" => FinishReason::ContentFilter,
        other => FinishReason::Unknown(other.to_string()),
    }
}

/// Collect the fields of `obj` that are not in `modeled` for `extra` carriage.
fn unmodeled_fields(
    obj: &serde_json::Map<String, Json>,
    modeled: &[&str],
) -> serde_json::Map<String, Json> {
    obj.iter()
        .filter(|(key, _)| !modeled.contains(&key.as_str()))
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect()
}

fn native_component(value: &Json) -> ProviderNativeComponent {
    ProviderNativeComponent {
        provider: "oci_genai".to_string(),
        kind: value
            .get("type")
            .and_then(Json::as_str)
            .unwrap_or("unknown")
            .to_string(),
        value: value.clone(),
    }
}

// ---------------------------------------------------------------------------
// GENERIC content conversion
// ---------------------------------------------------------------------------

/// Flatten a GENERIC content value into normalized [`MessageContent`].
///
/// A content-part list whose parts are all `{"type": "TEXT", "text": ...}` is
/// flattened to plain text; lists carrying any non-text part are preserved as
/// typed parts so image or future block types survive losslessly.
fn decode_generic_content(value: Option<&Json>) -> Result<Option<MessageContent>> {
    let value = match value {
        None | Some(Json::Null) => return Ok(None),
        Some(value) => value,
    };
    if let Some(text) = value.as_str() {
        return Ok(Some(MessageContent::Text(text.to_string())));
    }
    let parts = value.as_array().ok_or_else(|| {
        FlowError::InvalidArgument(
            "OCI GenAI GENERIC message content must be a string, an array, or null".into(),
        )
    })?;
    if parts.is_empty() {
        // Tool-call-only messages carry `"content": []`; there is no content.
        return Ok(None);
    }
    if let Some(text) = flatten_all_text_parts(parts) {
        return Ok(Some(MessageContent::Text(text)));
    }
    let parts = parts
        .iter()
        .map(decode_generic_content_part)
        .collect::<Result<Vec<_>>>()?;
    Ok(Some(MessageContent::Parts(parts)))
}

/// Join a part list into plain text when every part is a `TEXT` part.
fn flatten_all_text_parts(parts: &[Json]) -> Option<String> {
    let mut text = String::new();
    for part in parts {
        let obj = part.as_object()?;
        if obj.get("type").and_then(Json::as_str) != Some("TEXT") {
            return None;
        }
        match obj.get("text") {
            None | Some(Json::Null) => {}
            Some(Json::String(part_text)) => text.push_str(part_text),
            Some(_) => return None,
        }
    }
    Some(text)
}

fn decode_generic_content_part(value: &Json) -> Result<ContentPart> {
    let Some(obj) = value.as_object() else {
        return Err(FlowError::InvalidArgument(
            "OCI GenAI GENERIC content part must be an object".into(),
        ));
    };
    match obj.get("type").and_then(Json::as_str) {
        Some("TEXT") => Ok(ContentPart::Text {
            text: obj
                .get("text")
                .and_then(Json::as_str)
                .unwrap_or_default()
                .to_string(),
            extra: obj
                .iter()
                .filter(|(key, _)| !matches!(key.as_str(), "type" | "text"))
                .map(|(key, value)| (key.clone(), value.clone()))
                .collect(),
        }),
        _ => {
            let native = native_component(value);
            Ok(ContentPart::ProviderNative {
                provider: native.provider,
                kind: native.kind,
                value: native.value,
            })
        }
    }
}

// ---------------------------------------------------------------------------
// LlmResponseCodec implementation
// ---------------------------------------------------------------------------

impl LlmResponseCodec for OCIGenAIChatCodec {
    fn decode_response(&self, response: &Json) -> Result<AnnotatedLlmResponse> {
        let Some(obj) = response.as_object() else {
            // Non-object responses are preserved raw so observability still
            // captures whatever the provider path produced.
            let mut extra = serde_json::Map::new();
            extra.insert("raw".to_string(), response.clone());
            return Ok(AnnotatedLlmResponse {
                extra,
                ..AnnotatedLlmResponse::default()
            });
        };

        let (envelope, chat_response) = match obj.get("chatResponse").and_then(Json::as_object) {
            Some(chat_response) => (Some(obj), chat_response),
            None => (None, obj),
        };

        let model = envelope
            .and_then(|envelope| envelope.get("modelId"))
            .and_then(Json::as_str)
            .map(str::to_string);
        let model_version = envelope
            .and_then(|envelope| envelope.get("modelVersion"))
            .and_then(Json::as_str)
            .map(str::to_string);
        let api_format = chat_response
            .get("apiFormat")
            .and_then(Json::as_str)
            .unwrap_or("GENERIC")
            .to_uppercase();

        let (message, tool_calls, finish_reason, nested_extra) = match api_format.as_str() {
            "COHERE" => decode_cohere_response_body(chat_response),
            "COHEREV2" => decode_cohere_v2_response_body(chat_response)?,
            _ => decode_generic_response_body(chat_response)?,
        };

        let id = if api_format == "COHEREV2" {
            chat_response
                .get("id")
                .and_then(Json::as_str)
                .map(str::to_string)
        } else {
            None
        };

        let usage = chat_response
            .get("usage")
            .and_then(Json::as_object)
            .map(decode_oci_usage);

        // Preserve fields the normalized shape does not model so observability
        // keeps timeCreated, service tiers, grounding metadata, and future
        // provider fields.
        let modeled_response_keys: &[&str] = match api_format.as_str() {
            "COHERE" => &["apiFormat", "text", "finishReason", "toolCalls", "usage"],
            "COHEREV2" => &["apiFormat", "id", "message", "finishReason", "usage"],
            _ => &["apiFormat", "choices", "usage"],
        };
        let mut extra = match envelope {
            Some(envelope) => {
                unmodeled_fields(envelope, &["chatResponse", "modelId", "modelVersion"])
            }
            None => serde_json::Map::new(),
        };
        extra.extend(unmodeled_fields(chat_response, modeled_response_keys));
        extra.extend(nested_extra);

        Ok(AnnotatedLlmResponse {
            id,
            model,
            message,
            tool_calls,
            finish_reason: finish_reason.as_deref().map(map_oci_finish_reason),
            usage,
            optimization_summary: None,
            api_specific: Some(ApiSpecificResponse::OCIGenAI {
                api_format: Some(api_format),
                model_version,
            }),
            extra,
        })
    }
}

type ResponseBody = (
    Option<MessageContent>,
    Option<Vec<ResponseToolCall>>,
    Option<String>,
    serde_json::Map<String, Json>,
);

/// Keys of the decoded GENERIC choice consumed by the normalized shape.
///
/// `index` is excluded from `extra` carriage as well: it is positional trivia
/// (always `0` for the single decoded choice) rather than provider data.
const MODELED_CHOICE_KEYS: &[&str] = &["message", "finishReason", "index"];

/// Keys of a decoded assistant message consumed by the normalized shape.
const MODELED_MESSAGE_KEYS: &[&str] = &["role", "content", "toolCalls"];

/// Namespace unmodeled fields of a decoded nested container under `key`.
///
/// The choice-level fields of GENERIC responses (`logprobs`, `usage`,
/// `groundingMetadata`, `serviceTier`) and the message-level fields of
/// GENERIC (`refusal`, `annotations`, `reasoningContent`) and COHEREV2
/// (`toolPlan`, `citations`) responses are documented in the OCI schema but
/// not normalized; they are carried in `extra` under the container's wire key
/// so their origin stays unambiguous.
fn nest_unmodeled_fields(
    extra: &mut serde_json::Map<String, Json>,
    key: &str,
    obj: &serde_json::Map<String, Json>,
    modeled: &[&str],
) {
    let unmodeled = unmodeled_fields(obj, modeled);
    if !unmodeled.is_empty() {
        extra.insert(key.to_string(), Json::Object(unmodeled));
    }
}

fn decode_generic_response_body(
    chat_response: &serde_json::Map<String, Json>,
) -> Result<ResponseBody> {
    let mut nested_extra = serde_json::Map::new();
    let Some(first_choice) = chat_response
        .get("choices")
        .and_then(Json::as_array)
        .and_then(|choices| choices.first())
        .and_then(Json::as_object)
    else {
        return Ok((None, None, None, nested_extra));
    };
    nest_unmodeled_fields(
        &mut nested_extra,
        "choice",
        first_choice,
        MODELED_CHOICE_KEYS,
    );
    let finish_reason = first_choice
        .get("finishReason")
        .and_then(Json::as_str)
        .map(str::to_string);
    let Some(raw_message) = first_choice.get("message").and_then(Json::as_object) else {
        return Ok((None, None, finish_reason, nested_extra));
    };
    nest_unmodeled_fields(
        &mut nested_extra,
        "message",
        raw_message,
        MODELED_MESSAGE_KEYS,
    );
    let message = decode_generic_content(raw_message.get("content"))?;
    let tool_calls = raw_message
        .get("toolCalls")
        .and_then(Json::as_array)
        .map(|calls| decode_response_tool_calls(calls))
        .filter(|calls: &Vec<ResponseToolCall>| !calls.is_empty());
    Ok((message, tool_calls, finish_reason, nested_extra))
}

fn decode_cohere_response_body(chat_response: &serde_json::Map<String, Json>) -> ResponseBody {
    let message = chat_response
        .get("text")
        .and_then(Json::as_str)
        .map(|text| MessageContent::Text(text.to_string()));
    let tool_calls = chat_response
        .get("toolCalls")
        .and_then(Json::as_array)
        .map(|calls| decode_response_tool_calls(calls))
        .filter(|calls| !calls.is_empty());
    let finish_reason = chat_response
        .get("finishReason")
        .and_then(Json::as_str)
        .map(str::to_string);
    // COHERE (v1) is flat: unmodeled fields live directly on the chat
    // response and are already carried by the chat-response-level pass.
    (message, tool_calls, finish_reason, serde_json::Map::new())
}

/// Decode a COHEREV2 (`CohereChatResponseV2`) body: a single assistant
/// `message` whose `content` is a typed part list (`TEXT`, `THINKING`,
/// `IMAGE_URL`, `DOCUMENT`) and whose tool calls nest an OpenAI-style
/// `function` object.
fn decode_cohere_v2_response_body(
    chat_response: &serde_json::Map<String, Json>,
) -> Result<ResponseBody> {
    let mut nested_extra = serde_json::Map::new();
    let finish_reason = chat_response
        .get("finishReason")
        .and_then(Json::as_str)
        .map(str::to_string);
    let Some(raw_message) = chat_response.get("message").and_then(Json::as_object) else {
        return Ok((None, None, finish_reason, nested_extra));
    };
    nest_unmodeled_fields(
        &mut nested_extra,
        "message",
        raw_message,
        MODELED_MESSAGE_KEYS,
    );
    let message = decode_generic_content(raw_message.get("content"))?;
    let tool_calls = raw_message
        .get("toolCalls")
        .and_then(Json::as_array)
        .map(|calls| decode_response_tool_calls(calls))
        .filter(|calls: &Vec<ResponseToolCall>| !calls.is_empty());
    Ok((message, tool_calls, finish_reason, nested_extra))
}

/// Convert an OCI response tool-call list into [`ResponseToolCall`]s.
fn decode_response_tool_calls(calls: &[Json]) -> Vec<ResponseToolCall> {
    calls
        .iter()
        .enumerate()
        .filter_map(|(index, call)| decode_response_tool_call(index, call))
        .collect()
}

/// Convert an OCI response tool call into [`ResponseToolCall`].
///
/// GENERIC calls are flat (`{id, type, name, arguments}`) with `arguments` as a
/// JSON-encoded string; COHERE calls carry `name` plus parsed `parameters` and
/// no `id`, so a positional `call_{index}` id is synthesized to keep parallel
/// calls distinguishable; COHEREV2 calls nest `name`/`arguments` under an
/// OpenAI-style `function` object next to the `id`.
fn decode_response_tool_call(index: usize, value: &Json) -> Option<ResponseToolCall> {
    let obj = value.as_object()?;
    let body = obj.get("function").and_then(Json::as_object).unwrap_or(obj);
    let name = body.get("name")?.as_str()?.to_string();
    let arguments = match body.get("arguments") {
        Some(Json::String(text)) => {
            // CRITICAL: GENERIC arguments arrive JSON-encoded; parse for the
            // normalized shape, preserving the raw string when unparseable.
            serde_json::from_str::<Json>(text).unwrap_or_else(|_| Json::String(text.clone()))
        }
        Some(other) => other.clone(),
        None => body.get("parameters").cloned().unwrap_or(Json::Null),
    };
    let id = match obj.get("id").and_then(Json::as_str) {
        Some(id) => id.to_string(),
        None => format!("call_{index}"),
    };
    Some(ResponseToolCall {
        id,
        name,
        arguments,
    })
}

/// Map OCI usage counters onto the normalized [`Usage`] field names.
///
/// OpenAI and xAI models report cache hits under
/// `promptTokensDetails.cachedTokens`.
fn decode_oci_usage(usage: &serde_json::Map<String, Json>) -> Usage {
    let cache_read_tokens = usage
        .get("promptTokensDetails")
        .and_then(Json::as_object)
        .and_then(|details| details.get("cachedTokens"))
        .and_then(Json::as_u64);
    Usage {
        prompt_tokens: usage.get("promptTokens").and_then(Json::as_u64),
        completion_tokens: usage.get("completionTokens").and_then(Json::as_u64),
        total_tokens: usage.get("totalTokens").and_then(Json::as_u64),
        cache_read_tokens,
        cache_write_tokens: None,
        cost: None,
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
#[path = "../../tests/unit/codec/oci_genai_tests.rs"]
mod tests;
